lib/matplotlib/_constrained_layout.py

```diff
@@ -325,6 +327,17 @@ def get_margin_from_padding(obj, *, w_pad=0, h_pad=0,
               'bottomcb': h_pad, 'topcb': h_pad,
               'left': 0, 'right': 0,
               'top': 0, 'bottom': 0}
+
+    if isinstance(gs, mgridspec.GridSpecFromSubplotSpec):
```

Comment:
As noted in the issue, I'm not in favour of this change.  margins are to be relative to their outer container.  If the outer container also has a margin that should be respected as well.  

I understand aesthetically what you are trying to do.  However, you are basically overriding the point of a margin, which other users may have set this way on purpose.  Secondly, in a practical example with labels these axes will almost certainly not be lined up the way you are suggesting anyways.

Reply:
I've come up with an alternative solution to (hopefully) address these issues and opened a new PR: #28903 

See images in that PR which demonstrate testing with randomly sized labels, both before and after the new fix is applied.